### PR TITLE
Support validations for component attributes and elements

### PR DESCRIPTION
**Pull request by msalzburg**
We used the components gem in one of our recent web projects and really enjoyed it :)

To ensure the correct usage of our components we decided to "integrate" rails validations.
Now invalid components will raise validations errors which can be detected by feature specs.

This behaviour is currently enabled for all environments, probably it should be configurable for each environment.

Obviously this is still work in progress - I am just wondering if you would consider merging it?

### CODE REVIEW

**Comment by jensljungblad** [comment]
Good to hear you've found it enjoyable :)

Yes I think validations are an excellent addition. I had a `required` option in the past that could be passed to attributes, but I think leveraging Rails validations is a better idea, since you can validate so much more than just presence.

As for how to do it.. Leaving it up to the user to implement in their `ApplicationComponent` initially sounded pretty good to me. It could have been documented similar to what you have in this PR and looked something like:

```ruby
class Component < Components::Component
  include ActiveModel::Validations

  def initialize
    super
    validate!
  end
```

However, this only allows for validation of component attributes, and not of nested element attributes.

So that leaves either a) making it possible for the user to easily extend not only the base component, but the base element or b) include validations by default on `Components::Element`.

Have you thought of any issues with always including validations? There is already a dependency on Rails, so I wouldn't mind leveraging parts of `ActiveModel`, if it's for a good reason.

**Comment by imathis** [comment]
I've been working on implementing Components as well. I really like what you're doing here, but I too wish elements could extend other element classes. I was thinking about making a change to the elements class definition as follows:

```
```

**Comment by jensljungblad** [comment]
@imathis Seems like a nice solution! This is mostly to avoid having to define methods etc in the block?

**Comment by imathis** [comment]
@jensljungblad Having to treat each component and its elements as new creations doesn't work with my use case as well, so I'm working on making it possible to extend Elements and Components, allowing them to inherit elements, attributes, and methods, as well as allowing elements to render from templates. I'm not sure if that's something you're interested in or not, but I'd be happy to show you what I'm working on once I have it up and running.

**Comment by msalzburg** [comment]
> Have you thought of any issues with always including validations? There is already a dependency on Rails, so I wouldn't mind leveraging parts of `ActiveModel`, if it's for a good reason.

I like your thinking! Including it by default is no big pain and it makes the experience way better.

**Comment by msalzburg** [comment]
@jensljungblad I updated the branch accordingly. It should be mergeable now?

**Comment by msalzburg** [comment]
@imathis since this PR focuses on general component validation and is potentially done, maybe we should open a separate branch/pr for your element enhancement?

File: README.md [hunk 1]
```diff
@@ -423,6 +423,24 @@ end
 <%= component "button", label: "Sign in", url: sign_in_path %>
 ```
 
+### Validation
```
Comment (jensljungblad) [inline README.md]: Perhaps name "Attribute validations" and move below the "Attribute overrides" section?

File: README.md [hunk 2]
```diff
@@ -423,6 +423,24 @@ end
...
+
+To ensure your components get initialized properly you can use ActiveModel::Validations in your elements or components:
```
Comment (jensljungblad) [inline README.md]: Perhaps backticks around `ActiveModel::Validations` (and `ActiveModel::ValidationError` further down)

File: README.md [hunk 3]
```diff
@@ -423,6 +423,24 @@ end
...
+
+```ruby
+# app/components/button_component.rb %>
+
+class ButtonComponent < Components::Component
```
Comment (jensljungblad) [inline README.md]: And if we move this section up, perhaps use the `AlertComponent` for the example instead.

File: lib/components/element.rb [hunk 4]
```diff
@@ -45,16 +47,25 @@ def self.element(name, multiple: false, &config)
         get_instance_variable(plural_name)
       end
     end
+
     # rubocop:enable Metrics/AbcSize
     # rubocop:enable Metrics/CyclomaticComplexity
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/PerceivedComplexity
 
+    attr_reader :nested_block
```
Comment (jensljungblad) [inline lib/components/element.rb]: What is this for?

File: test/component_test.rb [hunk 5]
```diff
@@ -114,6 +116,30 @@ class ComponentTest < ActiveSupport::TestCase
     assert_equal [], component.foos
   end
 
+  class TestValidationComponent < Components::Component
+    attribute :foo
+    validates :foo, presence: true
+  end
```
Comment (jensljungblad) [inline test/component_test.rb]: I've been sticking the setup in each test case for now, just something I'm trying out :) Let's stick with that for now! So basically just move this into the test case, check the other test cases to see how.

File: lib/components/element.rb (hunk 4 above)
Comment (msalzburg) [reply lib/components/element.rb]: I introduced the nested_block attribute to be able to access the optional block during validation. This way validations can be conditional whether a block was given or not.

File: README.md (hunk 1 above)
Comment (msalzburg) [reply README.md]: Sure - will do.

File: README.md (hunk 2 above)
Comment (msalzburg) [reply README.md]: Sure - will do.

File: README.md (hunk 3 above)
Comment (msalzburg) [reply README.md]: np

File: test/component_test.rb (hunk 5 above)
Comment (msalzburg) [reply test/component_test.rb]: Primary I defined those classes separately to DRY up the code and dodge an active model validation issue with missing class names on anonymous classes. Since I managed to find a workaround for the missing names I used anonymous classes inside the tests.

File: lib/components/element.rb (hunk 4 above)
Comment (msalzburg) [reply lib/components/element.rb]: Any further thoughts?

File: test/component_test.rb [hunk 6]
```diff
@@ -114,6 +116,40 @@ class ComponentTest < ActiveSupport::TestCase
...
+
+  test "initialize without attribute and failing validation" do
+    component_class = Class.new(Components::Component) do
+      attribute :foo
+      validates :foo, presence: true
+
+      # supplies a name argument for active validations on anonymous classes
+      def self.model_name
+        ActiveModel::Name.new(self, nil, "temp")
+      end
```
Comment (jensljungblad) [inline test/component_test.rb]: Will this be a problem when putting validations on element attributes as well? Since they are anonymous classes. Could we have a test case for this as well? Something like:

```ruby
component_class = Class.new(Components::Component) do
  element :foo do
    attribute :bar
    validates :bar, presence: true
  end
end
```

File: test/component_test.rb [hunk 7]
```diff
@@ -114,6 +116,40 @@ class ComponentTest < ActiveSupport::TestCase
...
+    end
+
+    exception = assert_raises { component_class.new(:view) }
+    assert_instance_of ActiveModel::ValidationError, exception
```
Comment (jensljungblad) [inline test/component_test.rb]: You can assert the class with `assert_raises` I think: `assert_raises ActiveModel::ValidationError { component_class.new(:view) }`.

File: lib/components/element.rb (hunk 4 above)
Comment (jensljungblad) [reply lib/components/element.rb]: Oh I see, that makes sense I think. And you can also validate the presence of a block? Let me think about the naming of the variable though, I struggled with that one myself when I first wrote this :)

Comment (msalzburg) [reply lib/components/element.rb]: Currently there are two ways to interact with the given block:

1) You can access the block directly through the attribute reader "#nested_block"
2) You can check for its presence with the "#nested_block?" method

I was struggling with its naming too - any suggestions are very welcome ;)

File: test/component_test.rb (hunk 7 above)
Comment (msalzburg) [reply test/component_test.rb]: It takes a pair of brackets - but it works fine:

`assert_raises(ActiveModel::ValidationError) { component_class.new(:view)`

File: test/component_test.rb (hunk 6 above)
Comment (msalzburg) [reply test/component_test.rb]: Good point.

Comment (msalzburg) [reply test/component_test.rb]: I defined an ActiveModel::Name for the element and component class, which enables validations on anonymous class initialization.

**Comment by msalzburg** [comment]
Except of the "nested_block" naming issue all comments should be resolved. How should we proceed?

File: lib/components/component.rb [hunk 8]
```diff
@@ -1,5 +1,9 @@
 module Components
   class Component < Element
+    def self.model_name
+      ActiveModel::Name.new(Components::Component)
+    end
```
Comment (jensljungblad) [inline lib/components/component.rb]: Since `Component` inherits from `Element` it should be enough to define this on `Element`?

File: test/component_test.rb [hunk 9]
```diff
@@ -102,18 +104,82 @@ class ComponentTest < ActiveSupport::TestCase
...
+      validates :foo, presence: true
+    end
+
+    assert_nothing_raised { component_class.new(view_class.new) }
+  end
+
+  test "initialize element and successfull validation" do
+    component_class = Class.new(Components::Component) do
+      element :foo do
+        attribute :label
```
Comment (jensljungblad) [inline test/component_test.rb]: Not used, so can be removed

File: test/component_test.rb [hunk 10]
```diff
@@ -102,18 +104,82 @@ class ComponentTest < ActiveSupport::TestCase
...
+        attribute :bar
+
+        validates :bar, presence: true
+      end
+    end
+
+    e = assert_raises(ActiveModel::ValidationError) do
+      component_class.new(view_class.new, {}) do |c|
+        c.foo(label: "lalal") { "label" }
```
Comment (jensljungblad) [inline test/component_test.rb]: Perhaps no need to have a `label` attribute here either.

File: test/component_test.rb [hunk 11]
```diff
@@ -5,35 +5,37 @@ class ComponentTest < ActiveSupport::TestCase
...
     assert_equal "foo", component.to_s
+    assert_equal true, component.nested_block?
   end
 
   test "initialize attribute with no value" do
     component_class = Class.new(Components::Component) do
       attribute :foo
     end
-    component = component_class.new(:view)
+    component = component_class.new(view_class.new)
```
Comment (jensljungblad) [inline test/component_test.rb]: Good call!

File: lib/components/element.rb (hunk 4 above)
Comment (jensljungblad) [reply lib/components/element.rb]: Aha, I basically thought you just renamed the `@yield` instance variable, but now I see that's not the case.

When do you need to access the block and not the captured content of the block? I.e. when do you need `nested_block` instead of what's in `@yield`?

Comment (jensljungblad) [reply lib/components/element.rb]: Let me rephrase.. Do we need the additional instance variable? I.e. both `@yield` and `@nested_block`, or would one suffice?

File: lib/components/component.rb (hunk 8 above)
Comment (msalzburg) [reply lib/components/component.rb]: Technically it would work without redefining the model_name on the `Component` class, but then anonymous instances of the `Component` would have the `Components::Element` model name instead of `Components::Component`. Guess it is more like a flavor call. I would prefer to redefine it, but if you dont I can remove it?

Comment (jensljungblad) [reply lib/components/component.rb]: Oh you're absolutely right, I didn't see the difference. Let's keep it like this!

File: test/component_test.rb (hunk 9 above)
Comment (msalzburg) [reply test/component_test.rb]: Done.

File: test/component_test.rb (hunk 10 above)
Comment (msalzburg) [reply test/component_test.rb]: Removed label attribute here too-

File: lib/components/element.rb (hunk 4 above)
Comment (msalzburg) [reply lib/components/element.rb]: I guess we can eliminate one instance variable. Until now it was only required to validate for the presence of the block via the `nested_block?` (or how it will be named) method. But lets work out the naming of the instance variable and the xxx? method first. Any naming suggestions/preferences?

Comment (jensljungblad) [reply lib/components/element.rb]: I think it depends on what we want to be able to do. If we only want to support conditional validations depending on if a block was given or not, we could just add:

```ruby
class Element
  def block_given? # or something
    @yield.present?
  end
end

class SomeComponent
  validates :something, if: :block_given?
end
```

And skip the reader. However, do we also want to support presence of the block? 

```ruby
class SomeComponent
  validates :block, presence: true
end
```

If so, we do need a reader, and a name for it. I don't think `yield` is very good since it's a reserved word in Ruby. One idea had before was `content` or some variation thereof, but I would prefer to use something that you wouldn't  also want to use as an attribute name. So maybe some variation on `block` is a good idea. Off the top of my head: 

- `content_block`
- `yielded_block`
- `captured_block`
- `block`

Comment (msalzburg) [reply lib/components/element.rb]: I just reviewed our use cases and they seem to match with your ideas quite nicely 👍 

We often used the basic presence validation on the `attr_reader`

```ruby
class SectionComponent < ApplicationComponent
  validates :content_block, presence: true
end
```
and sometimes had more "complex" validations using the `content_block?` method

```ruby
class OtherComponent < ApplicationComponent
  attribute :text

  validate :validate_content_is_present

  def validate_content_is_present
    return if content_block? || text.present?

    errors.add(:base, 'no block or text attribute is given!')
  end
end
```
Also the `content_block?`method was used in several component view templates to modify their rendering.

Guess we implement both then?

BTW I think naming it `content_block` really is an improvement. Nice one!

**Comment by msalzburg** [comment]
Everything fine?

File: lib/components/element.rb [hunk 12]
```diff
@@ -45,20 +51,28 @@ def self.element(name, multiple: false, &config)
         get_instance_variable(plural_name)
       end
     end
+
     # rubocop:enable Metrics/AbcSize
     # rubocop:enable Metrics/CyclomaticComplexity
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/PerceivedComplexity
 
+    attr_reader :content_block
```
Comment (jensljungblad) [inline lib/components/element.rb]: Should we document this as well as `content_block?` somewhere? I'm not sure if it belongs under the validation section or perhaps earlier, under "Attributes and blocks". I suppose they could be used for other things than validation. What do you think?

Do you think `block_content` is clearer than `content_block` btw? As in, "the content of the block", since the variable doesn't actually contain a block, but rather the yielded content of the block.

Comment (msalzburg) [reply lib/components/element.rb]: 👍 on `block_content` - I already renamed it.

I was wondering if we should use the 'block_content' reader as preferred way of rendering blocks in view templates - instead of rendering the component instance via the to_s method. It clarifies its output and aligns with other methods and attribute calls we make on the instance (e.g. alert.context or alert.block_content?). In this case we should probably deprecate the to_s method?

```ruby
<% # app/components/alert/_alert.html.erb %>

<div class="alert alert--<%= alert.context %>" role="alert">
  <%= alert.block_content %>
</div>
```

instead of

```ruby
<% # app/components/alert/_alert.html.erb %>

<div class="alert alert--<%= alert.context %>" role="alert">
  <%= alert %>
</div>
```

Comment (imathis) [reply lib/components/element.rb]: I have a strong interest in moving away from `to_s` as well, although I would love to see elements either dump their content or support a render method so that they can specify their own templating if need be.

Comment (jensljungblad) [reply lib/components/element.rb]: It's a valid question for sure. The reason I did it in the first place was because I couldn't come up with a good name, that wouldn't potentially clash with an attribute, and that felt short enough to type over and over in a lot of components.

Perhaps let's keep `to_s` for now, to remain backwards compatible. At least there is now an alternative with `block_content`.

@imathis Probably a separate discussion, but what would be the reason for specifying templates for elements? You could always use a separate component, but perhaps you still want them to be internal to the component, and not usable from outside? I suppose you could always implement this by defining a helper:

```ruby
class Component
  element :body

  def render_body
    @view.render(partial: "../path/to/body/partial", object: body)
  end
end

Comment (msalzburg) [reply lib/components/element.rb]: @jensljungblad should I update the README to use `alert.content_block` als default then and mention the "instance to_s" rendering as "shortcut"? We could also not mention the "instance to_s" rendering, still support it and add a deprecation warning via log file? Describing two way to render the components block seems a bit awkward to me ...

Comment (jensljungblad) [reply lib/components/element.rb]: I guess I don't really care about backwards compatibility at this point. It's still not even a released gem :) Sure, let's remove `#to_s` and update the README to use `#block_content` instead. More verbose for sure but perhaps makes more sense. And can always be revisited down the line if someone has a better idea.

Comment (msalzburg) [reply lib/components/element.rb]: After eliminating the to_s method and using the block_content method for block rendering instead, I get your point on verbosity. Especially the rendering of "simple" elements (without further attributes) appears  a bit bulky (see changes on component views in dummy/app)

Maybe a shorter name is appropriate. We could just call it `block` instead of `block_content` (and `block?` instead of `block_content?`)?

Comment (jensljungblad) [reply lib/components/element.rb]: Or maybe `content` isn't a terrible name. It's close to Rails own `content_for` helpers, which do a similar thing: https://api.rubyonrails.org/classes/ActionView/Helpers/CaptureHelper.html#method-i-capture. I suppose I ended up using `to_s` so I wouldn't have to name this thing :)

I suppose it could be a problem if you have an attribute named `content`. But maybe that's not a big deal.

Another alternative is to merge validations, and create a new issue for adding a reader/moving away from `to_s`. You could implement the `block_content?` method yourself in the meanwhile:

```ruby
class Component
  def block_content?
    @yield.present?
  end
end
```

Comment (imathis) [reply lib/components/element.rb]: I will say I nearly used `content` as an attribute for an attribute once, and I can imagine it's quite a reasonable choice. Having to share space with an end-user's attributes and elements puts any choice in danger of breaking things. I've been working on a PR which adds features to elements, but fear of breaking things due to naming conflicts makes it hard to expect upstream adoption. Is there a way to protect collisions for internal instance methods or attributes? Like (forgive me) underscoring internal methods, e.g., `_content`.

Comment (msalzburg) [reply lib/components/element.rb]: I just reviewed our project again and noticed that we have been building several components with an `content` attribute by instinct 🤓

Comment (msalzburg) [reply lib/components/element.rb]: Both names have their pros and cons. From my perspective some content gets injected into the component by a ruby block, therefore it somehow feels natural to reference that block by calling `block`. 

@jensljungblad I guess the final naming is your call ;) Feel free to rename as you like and lets 🚀

Comment (jensljungblad) [reply lib/components/element.rb]: Let's extract this discussion from this PR, it feels like a bit of an aside, and not crucial to validations. I created an issue here https://github.com/jensljungblad/components/issues/32 where we can discuss more about the block and its API.

If you want we can just remove the block stuff from this PR and merge it, and we'll have validations for attributes ready for now.

**Comment by jensljungblad** [comment]
Do we want to support validating the presence of elements? Is that a use case you've come across? I suppose then that validation would have to be moved to the render stage, since the elements are added after initialization.

File: lib/components/element.rb (hunk 12 above)
Comment (msalzburg) [reply lib/components/element.rb]: All changes concerning the block handling have been reverted.

**Comment by msalzburg** [comment]
Element presence validation already works ;) But I also added some tests.

Guess this branch is finally mergeable and we will continue in #32

**Comment by jensljungblad** [comment]
@msalzburg Haha right, because we yield the block before we run `validate!`.. Excellent :)

**Comment by jensljungblad** [comment]
@msalzburg Thanks for the contribution, great work!
